[Inductor][Triton][FP8] Refactor scaled_mm template to accept scaling mode

jananisriram opened the pull request:
Summary: Refactor `scaled_mm` Inductor template to support template choice based on scaling mode. This modification sets up the infrastructure for adding new templates based on new scaling modes, such as deepseek-style scaling (a follow-up diff), as new scaling modes (deepseek, block, group) scale before the accumulation (as opposed to per-tensor and per-row scaling, which apply scaling after accumulation). This modification also further enables Inductor to infer a scaling type based on the shape of the scaling tensors, which makes existing infrastructure more extensible to new scaling modes.

Test Plan:
```
TORCHINDUCTOR_CACHE_DIR=~/personal/cache_dir_inductor CUDA_LAUNCH_BLOCKING=1 TORCH_USE_CUDA_DSA=1 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 ENABLE_PERSISTENT_TMA_MATMUL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 buck2 run mode/{opt,inplace} pytorch/tritonbench:run -- --op fp8_gemm --only torch_fp8_gemm,pt2_fp8_gemm --metrics tflops,accuracy --m 256 --n 768 --k 512 --output="/home/jananisriram/personal/random_bench.csv" --scaling_rowwise --atol=20 --rtol=2 2>&1 | tee ~/personal/random.log
```

Differential Revision: D83591083


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

PaulZhang12 commented on torch/_inductor/kernel/mm.py:
```diff
@@ -421,7 +421,7 @@ def apply_scaling(
     stride_bn = {{stride("B", 1)}}
 
-    if SCALING_ROWWISE:
+    if SCALING_MODE == 1:
```
What is `SCALING_MODE == 1` indicate here?

jananisriram commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1341,8 +1341,27 @@ def tuned_scaled_mm(
+            # Supported scaling modes (from aten/src/ATen/native/cuda/Blas.cpp::1163):
+            # TensorWise, RowWise
+            # TODO (jananisriram): support BlockWise1x16, BlockWise1x32, BlockWise1x128, BlockWise128x128
+            if _is_scalar_like(scale_a_size) and _is_scalar_like(scale_b_size):
```
yeah this makes sense to me to avoid confusion

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -378,15 +380,15 @@ def apply_scaling(
     stride_b_scale_n,
 ):
-    if SCALING_ROWWISE:
+    if SCALING_MODE == 1:
```
can you pass Int enums into triton?

jananisriram replied:
afaik you can't (which is why you have to manually check == 1 here)

jananisriram replied:
just checked and it doesn't seem like we can!

slayton58 reviewed (approved):
Looks good, minor nit on enum

jananisriram commented:
@pytorchbot --help

jananisriram commented:
@pytorchbot rebase -b origin/main

jananisriram committed:
[Inductor][Triton][FP8] Refactor scaled_mm template to accept scaling mode (#164318)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/164318

Refactor `scaled_mm` Inductor template to support template choice based on scaling mode. This modification sets up the infrastructure for adding new templates based on new scaling modes, such as deepseek-style scaling (a follow-up diff), as new scaling modes (deepseek, block, group) scale before the accumulation (as opposed to per-tensor and per-row scaling, which apply scaling after accumulation). This modification also further enables Inductor to infer a scaling type based on the shape of the scaling tensors, which makes existing infrastructure more extensible to new scaling modes.

Test Plan:
```
TORCHINDUCTOR_CACHE_DIR=~/personal/cache_dir_inductor CUDA_LAUNCH_BLOCKING=1 TORCH_USE_CUDA_DSA=1 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 ENABLE_PERSISTENT_TMA_MATMUL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 buck2 run mode/{opt,inplace} pytorch/tritonbench:run -- --op fp8_gemm --only torch_fp8_gemm,pt2_fp8_gemm --metrics tflops,accuracy --m 256 --n 768 --k 512 --output="/home/jananisriram/personal/random_bench.csv" --scaling_rowwise --atol=20 --rtol=2 2>&1 | tee ~/personal/random.log
```

Reviewed By: slayton58

Differential Revision: D83591083

seemethere commented:
@pytorchbot merge -f 'already merged internally'
